src/shared/vscode/commands2.ts

```diff
@@ -0,0 +1,330 @@
+
+        if (registered?.declared) {
+            throw new Error(`Command "${resource.id}" has already been declared by the Toolkit`)
+        }
+
+        this.resources.set(resource.id, resource)
+        return resource
+    }
+
+    static #instance: Commands
+
+    public static get instance(): Commands {
+        return (this.#instance ??= new Commands())
+    }
+}
+
+/**
+ * Short-hand for {@link Commands.get}.
+ */
+export const getCommand = Commands.instance.get.bind(Commands.instance)
+
+/**
+ * Short-hand for {@link Commands.register}.
+ */
+export const registerCommand = Commands.instance.register.bind(Commands.instance)
+
+/**
+ * Short-hand for {@link Commands.declare}.
+ */
+export const declareCommand = Commands.instance.declare.bind(Commands.instance)
+
+/**
+ * Short-hand for {@link Commands.from}.
+ */
+export const fromClass = Commands.instance.from.bind(Commands.instance)
+
+type Functions<T> = { [P in keyof T]: T[P] extends Callback ? T[P] : never }
+type FunctionKeys<T> = { [P in keyof T]: T[P] extends Callback ? P : never }[keyof T]
+
+interface Declare<T, F extends Callback> {
+    (id: string): DeclaredCommand<F, [target: T]>
+}
+
+type Declarables<T> = {
+    [P in FunctionKeys<T> as `declare${Capitalize<P & string>}`]: Declare<T, Functions<T>[P]>
+}
+
+/**
+ * Returns all functions found on the target's prototype chain.
+ *
+ * Conflicts from functions sharing the same key are resolved by order of appearance, earlier
+ * functions given precedence. This is equivalent to how the prototype chain is traversed when
+ * evaluating `target[key]`, so long as the property descriptor is not a 'getter' function.
+ */
+function getFunctions<T>(target: new (...args: any[]) => T): Functions<T> {
+    const result = {} as Functions<T>
+
+    for (const k of Object.getOwnPropertyNames(target.prototype)) {
+        if (typeof target.prototype[k] === 'function') {
+            result[k as keyof T] = target.prototype[k]
+        }
+    }
+
+    const next = Object.getPrototypeOf(target)
+    return next && next.prototype ? { ...getFunctions(next), ...result } : result
+}
+
+// TODO(sijaden): implement decoupled tree-view, then move this
+type ExcludedKeys = 'id' | 'label' | 'collapsibleState'
+interface LabeledTreeItem extends Omit<vscode.TreeItem, ExcludedKeys> {
+    readonly label: string
+}
+
+type PartialCommand = Omit<vscode.Command, 'arguments' | 'command'>
+type PartialTreeItem = Omit<LabeledTreeItem, 'command'>
+
+interface Builder {
+    asUri(): vscode.Uri
+    asCommand(content: PartialCommand): vscode.Command
+    asTreeNode(content: PartialTreeItem): AWSTreeNodeBase
+    asCodeLens(range: vscode.Range, content: PartialCommand): vscode.CodeLens
+}
+
+interface Deferred<T extends Callback, U extends any[]> {
+    readonly id: string
+    readonly name?: string
+    readonly factory: CommandFactory<T, U>
+}
+
+class CommandResource<T extends Callback = Callback, U extends any[] = any[]> {
```

Comment:
so `Commands` is like the control layer, and `CommandResource` is the command impl?  a docstring like that will be helpful.

```suggestion
/**
 * Contains the command implementation and properties. CommandResources are registered on the `{@link Commands}` singleton.
 */
class CommandResource<T extends Callback = Callback, U extends any[] = any[]> {
```

Reply:
Yeah `CommandResource` is just a helper class (that's why it's not exported). 

But I'll add a docstring.